feat: add basic gRPC health service

tustvold opened the pull request:
Implements the standard gRPC health service using [tonic-health](https://github.com/hyperium/tonic/tree/master/tonic-health) (part of tonic).

The implementation is extremely basic at the moment, but it should provide a starting point.

To test install grpc-health-probe from [here](https://github.com/grpc-ecosystem/grpc-health-probe/releases/tag/v0.3.6) and then run

```
$ grpc_health_probe-linux-amd64 -addr 127.0.0.1:8082 -service influxdata.platform.storage.Storage
status: SERVING
```

alamb reviewed (commented):
I would like to request two things:
1. Update the documentation / readme.md with the example of how to invoke a health check (e.g. copy/paste the info from your PR description)
2. Add some sort of test (perhaps something in the end-to-end test?)

The tests are to ensure that someone such as myself which doesn't know what they are doing accidentally break this health check service when the code is changed / refactored)

tustvold commented:
> perhaps something in the end-to-end test

Still working on how to do end-to-end testing of gRPC APIs. I'll put this in draft until such a time as I have the story for that fleshed out

tustvold commented:
I've updated the README and added an HTTP /health endpoint, but will need to include the grpc definitions to create a client for integration testing. Once #847 is in I should be able to move forward on this

tustvold commented on influxdb_iox_client/src/client/health.rs:
```diff
@@ -0,0 +1,66 @@
+
+impl Client {
+    /// Create a new client with the provided endpoint
+    pub async fn connect<D>(dst: D) -> Result<Self>
```
Currently this module provides a ClientBuilder and a FlightClientBuilder. As we add more gRPC services this is going to result in a lot of code duplication.

I'm therefore thinking of moving towards the following, with this PR being the first of many in this space:

* Expose the client modules, i.e. don't expose every different Client on the root module
* Allow creating clients from anything convertible into an Endpoint
* Add an EndpointBuilder that provides a generic way to set timeouts, etc... for an endpoint, that can then be shared across all gRPC client implementations

alamb replied:
What about moving towards having a single Client (for all of the various services, as they all are listening on the same underlying http port, right?)

tustvold replied:
I'd worry that the resulting client might become a touch unwieldy? Ultimately the API is split into different services to accommodate different use cases. A Flight client isn't going to be using the Management client endpoints, IOx clients will only use the not yet extant Replication API, etc... I admit it's a bit of a stylistic thing, but keeping separate concerns separate seems better to me?

alamb replied:
Yeah I agree it comes with tradeoffs -- a single client can more easily share common implementation / configuration items; If you have multiple clients you will have to have some way to share that commonality (which I what you are proposing).

I don't have strong feelings one way or the other

tustvold replied:
> The functionality already seems separated by the various Tonic generated traits, so having all the impls on a single Client doesn't feel too terrible to me

The clients don't implement the traits in most cases, as we want to provide our own error types, we could provide our own traits, but this feels a bit derived. I think I'm going to proceed with this approach and revisit if issues arise

alamb replied:
Sounds like a good plan to me

alamb commented on tests/end-to-end.rs:
```diff
@@ -396,7 +417,7 @@ impl TestServer {
 
         let try_http_connect = async {
             let client = reqwest::Client::new();
-            let url = format!("{}/ping", HTTP_BASE);
```
I think given that the `/ping` url is used to make sure the IOx service is up when running in k8s / idpe, I worry about losing test coverage for it in the iox end to end tests.

Do you think there is value in a specific test for /ping too?

alamb replied:
Or maybe even going futher and make `/ping` just return the same response as `/health` (as part of some future PR, perhaps)

tustvold replied:
I was not aware of this, I assumed `/ping` was just a random test endpoint... Using `/ping` as a healthcheck endpoint is not what I expected. Can/should we change this?

alamb replied:
I think switching to something more standard like `/health` and removing `/ping` sounds like a great idea. `/ping` got put in as a basic "we need something to work"

pauldix replied:
Agree, move to `/health`. Ping was a relic of InfluxDB from, I think, 2013. Let's lose that vestigial tail!

alamb reviewed (approved):
Looks great to me. Nice work @tustvold
